[fix/docs]: Improve CONTRIBUTING.md (#1300)
TheAlgorithms/C-Plus-Plus: CONTRIBUTING.md

# CONTRIBUTION GUIDELINES

## Before contributing

> 

Reviewer: These blank lines disappear when the markdown is rendered so they are not too vital but consistency is helpful.

Author: Yep, they do disappear. Anyways, this helps when
reading it in plain text (non-Markdown format). 🙂

It also helps the Markdown Lint; `CONTRIBUTING.md` had 10+ warnings.

Welcome to [TheAlgorithms/C-Plus-Plus](https://github.com/TheAlgorithms/C-Plus-Plus)! Before submitting pull requests, please make sure that you have **read the whole guidelines**. If you have any doubts about this contribution guide, please open [an issue](https://github.com/TheAlgorithms/C-Plus-Plus/issues/new/choose) and clearly state your concerns.

## Contributing

### Contributor

We are very happy that you consider implementing algorithms and data structures for others! This repository is referred to and used by learners from around the globe. Being one of our contributors, you agree and confirm that:

- You did your own work.

>     - No plagiarism allowed.  Any plagiarized work will not be merged.

Reviewer: I thought that markdown wanted four space indents.

Author: Nope, Visual Studio Code reports an error. It's only 2 if you want to make it look like:

- One indent
  - Two indents

Reviewer: Learn something new every day!

### Making Changes

#### Code

- Please use the directory structure of the repository.

> - File extension for code should be `*.hpp` `*.cpp`.

Reviewer 2: `*.h` is also fine; no need to remove that.

Author: AFAIK, if we use `.h` in C++, it will report an error.

Reviewer 2: are you sure? `https://github.com/TheAlgorithms/C-Plus-Plus/tree/master/numerical_methods` this folder has `.h` files and all is fine.

Author: Whoops, my bad. Added `*.h` format. 😄

- You can suggest reasonable changes to existing algorithms.
- Strictly use snake_case (underscore_separated) in filenames.
- If you have added or modified code, please make sure the code compiles before submitting.
- Our automated testing runs [__CMake__](https://cmake.org/) on all pull requests so please be sure that your code passes before submitting.
- Please conform to [doxygen](https://www.doxygen.nl/manual/docblocks.html) standard and document the code as much as possible. This not only facilitates the readers but also generates the correct info on website.
- **Be consistent in use of these guidelines.**

#### Documentation

- Make sure you put useful comments in your code.  Do not comment things that are obvious.
- Please avoid creating new directories if at all possible. Try to fit your work into the existing directory structure. If you want to create a new directory, then please check if a similar category has been recently suggested or created by other pull requests.
- If you have modified/added documentation, please ensure that your language is concise and contains no grammar errors.

> - Do not update `README.md` along with other changes, first create an issue and then link to that issue in your pull request to suggest specific changes required to README.md

Reviewer 2: ```suggestion
- Do not update [`README.md`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/README.md) along with other changes. First create an issue and then link to that issue in your pull request to suggest specific changes required to [`README.md`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/README.md).
```

- The repository follows [Doxygen](https://www.doxygen.nl/manual/docblocks.html) standards and auto-generates the [repository website](https://thealgorithms.github.io/C-Plus-Plus). Please ensure the code is documented in this structure. Sample implementation is given below.

#### Test

- Make sure to add examples and test cases in your main() function.
- If you find any algorithm or document without tests, please feel free to create a pull request or issue describing suggested changes.

> - Please try to add one or more `test()` functions that will invoke the algorithm implementation on random test data with expected output. Use `assert()` function to confirm that the tests will pass. Requires adding the `cassert` library.

Reviewer 2: ```suggestion
- Please try to add one or more `test()` functions that will invoke the algorithm implementation on random test data with expected output. Use `assert()` function to confirm that the tests will pass. Requires including the `cassert` header.
```

- Make sure to add examples and test cases in your main() function.
- If you find any algorithm or document without tests, please feel free to create a pull request or issue describing suggested changes.
- Please try to add one or more `test()` functions that will invoke the algorithm implementation on random test data with expected output. Use `assert()` function to confirm that the tests will pass.
- Please try to add one or more `test()` functions that will invoke the algorithm implementation on random test data with expected output. Use `assert()` function to confirm that the tests will pass. Requires adding the `cassert` library.

#### Typical structure of a program

#### Typical structure of a program:
```cpp
/**
 * @file 
 * @file
 * @brief Add one line description here
 * @details 
 * @details
 * This is a multi line
 * description containing links, references,
 * math equations, etc
 * @author [Name](https://github.com/handle)
 * @see related_file.cpp, another_file.cpp
 */

#include 
#include <cassert>
#include

/**
 * @namespace <check from other files in this repo>
 */
namespace name {

> 

Reviewer 2: keep this blank line

    return true;
}
} // namespace name

/**
 * @brief Test implementations
 * @returns void
 */
static void test() {
    /* desciptions of the following test */
    assert(func(...) == ...); // this ensures that the algorithm works as expected

    // can have multiple checks
}

/**
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
 * @returns 0 on exit

>  */

Reviewer 3: This sounds a bit too verbose to me. As a reader, I'm not learning anything new and therefore it beats the purpose of having the comment in the first place.

    return true;
}

> /** Test function */

Reviewer 2: missing closing parentheses for the namespace :)
```suggestion
} // namespace name
```

#### Building Locally

Before submitting a pull request,
 build the code locally or using the convenient [![Gitpod Ready-to-Code](https://img.shields.io/badge/Gitpod-Ready--to--Code-blue?logo=gitpod)](https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus) service.

```bash
cmake -B build -S .
```

#### Static Code Analyzer

We use [clang-tidy](https://clang.llvm.org/extra/clang-tidy/) as a static code analyzer with a configuration in [.clang-tidy](.clang-tidy).

```bash
clang-tidy --fix --quiet -p build subfolder/file_to_check.cpp --
```

#### Code Formatter

[__clang-format__](https://clang.llvm.org/docs/ClangFormat.html) is used for code forrmating.

- Installation (only needs to be installed once.)
  - Mac (using home-brew): `brew install clang-format`
  - Mac (using macports): `sudo port install clang-10 +analyzer`
  - Windows (MSYS2 64-bit): `pacman -S mingw-w64-x86_64-clang-tools-extra`
  - Linux (Debian): `sudo apt-get install clang-format-10 clang-tidy-10`
- Running (all platforms): `clang-format -i -style="file" my_file.cpp`

#### GitHub Actions

- Enable GitHub Actions on your fork of the repository.

> After enabling it will execute `clang-tidy` and `clang-format` after every a push (not a commit).

Reviewer 2: ```suggestion
After enabling it will execute `clang-tidy` and `clang-format` after every push (not a commit).
```
